### DISCUSSION TITLE
FileUpload custom file name

### DISCUSSION THREAD

**Discussion by tmrbaset** [filamentphp/filament#1729] [Help]
As far as I understood from the documentation, there is no method to give uploaded file a custom name, add prefix or suffix. It's either preserve the file name, or give it a random name. 

The problem with preserving file name, is that if this file name already exists, you wouldn't know, and I think the file is not saved.

So What is the workaround here? is there any method or workaround, to either define a custom name, or change the original file name by adding the date for example?

Thanks.

Note: I am not an expert in Livewire.

**Comment by titantwentyone** [comment]
I encountered this and came up with a solution which I need to properly formalize. First I added a macro to the Storage facade to generate a unique filename. This looks for any duplicates and adds a suffix if needed; incrementing it as needed.

```
Storage::macro('getUniqueFilename', function($disk, $originalFileName) {

    $filename = File::name($originalFileName);
    $extension = File::extension($originalFileName);

    if(Storage::disk($disk)->exists($originalFileName)) {
        
        $allFiles = collect(Storage::disk($disk)->allFiles())
            ->map(fn($file) => File::name($file)) //take the filenames only
            ->reject(fn($item) => $item == "") //reject empty filenames
            ->reject(fn($item) => !preg_match("/{$filename}-[0-9]*$/", $item)) //reject anything that doesnt match filename-xx
            ->sort(); 
            

        if($allFiles->count())
        {
            $f = $allFiles->last();
            $filename = ++$f;
        }
        else
        {
            //just add a suffix
            $filename .= "-01";
        }
    }

    $filename .= ".".$extension;

    return $filename;

});
```
For the Filament control, I modified how the image was uploaded:

```
FileUpload::make('image')
    ->saveUploadedFileUsing(function (FileUpload $component, TemporaryUploadedFile $file): string {

        $storeMethod = $component->getVisibility() === 'public' ? 'storePubliclyAs' : 'storeAs';

        $filename = str_replace(" ", "-", $file->getClientOriginalName());
        $filename = Storage::getUniqueFilename('public', $filename);

        return $file->{$storeMethod}($component->getDirectory(), $filename, $component->getDiskName());
    })
```
For extra points, I added a modification to remove the file (something the control doesn't do automatically):

```
FileUpload::make('image')
    ->deleteUploadedFileUsing(function($file) {

        Storage::disk('public')->delete($file);

    })
```
I actually wrapped all this up into a Trait and use it in my Forms. When I get time, I'll see if I can get this into a PR. Gist [here](https://gist.github.com/titantwentyone/99c7c678b690f10591daa64f95702b4b).

**Reply by tmrbaset** [reply]
This looks interesting, thanks @titantwentyone, I will give it a try.

**Reply by titantwentyone** [reply]
Let me know how you get on. May need some tweaking.

**Reply by danharrin** [reply]
This is interesting. Maybe we should implement a `getFileNameForStorageUsing(fn ($file) => $file->whatever)` method so you can implement this logic without having to override the `storage()` logic.

**Reply by danharrin** [reply]
If you'd like to have a crack at implementing that method in your PR, that would be great.

**Reply by titantwentyone** [reply]
Will do. Is the deletion okay? I can create a separate PR for that. Seems too simple (meaning I'm suspicious as to whether this is correct!) but it has worked for me.

**Reply by danharrin** [reply]
I made the deletion method very simple so you can delete files if you want, but I don't think it should be default behaviour, as uploaded files can often have multiple dependencies, and deleting a file from one record may have knock-on effects elsewhere. However, for the Spatie media library component, this is default behaviour since there is only ever one linked record for a file.

**Reply by titantwentyone** [reply]
No worries. Will just look at filenames.

**Reply by tmrbaset** [reply]
Thanks a lot @titantwentyone, this worked flawlessly, exactly what I needed.
I agree it should be added to the package, It will be very helpful for anyone using FileUpload.

**Reply by titantwentyone** [reply]
PR #1828 submitted! :)
